YLP-263 accept language for signup email

BigFab opened the pull request.

fabrice-diabeloop committed:
YLP-263 accept language for signup email

yvon-dblg reviewed (changes requested):
Missing reset-password

fabrice-diabeloop committed:
Add language in request for forgot password

ydrisR commented on CHANGELOG.md:
```diff
@@ -3,7 +3,9 @@
 This API sends notifications to users for things like forgotten passwords, initial signup, and invitations.
 
-## Unreleased
+## 1.2.3 - 2020-11-04
```
```suggestion
## 1.3.0 - 2020-11-04
```

This is a change

BigFab replied:
is ok now

yvon-dblg commented on api/signup.go:
```diff
@@ -163,10 +163,14 @@ func (a *Api) sendSignUpInformation(res http.ResponseWriter, req *http.Request,
 // @Description  forces the user to the confirmation-required page until the signup has been confirmed.
 // @Description  It sends an email that contains a random confirmation link
+// @Description  The email is sent in the language defined by "x-tidepool-language" header
+// @Description  Otherwise if null "Accept-Language" header otherwise if null "ENglish"
```
```suggestion
// @Description  Otherwise if null "Accept-Language" header otherwise if null "English"
```
?

BigFab replied:
this one is on purpose as we actually return "EN" for "English"

yvon-dblg commented on api/forgot.go:
```diff
@@ -49,10 +51,16 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
+	// even if not selected there is one by default so we should normally always end up with
+	// a value in GetUserChosenLanguage()
+	// however, just to play it safe, we can continue taking the browser preferred language
+	// or ENglish as default values
```
```suggestion
	// or English as default values
```
?

BigFab replied:
this one is on purpose as we actually return "EN" for "English"

yvon-dblg commented on api/forgot.go:
```diff
@@ -49,10 +51,16 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
+	// a value in GetUserChosenLanguage()
+	// however, just to play it safe, we can continue taking the browser preferred language
+	// or ENglish as default values
 	// In case the reseter is found a known user and has a language set, the language will be overriden in a later step
```
```suggestion
	// In case the resetter is found a known user and has a language set, the language will be overridden in a later step
```

yvon-dblg commented on api/forgot.go:
```diff
@@ -49,10 +51,16 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
 	var resetCnf *models.Confirmation
 	var reseterLanguage string
```
```suggestion
	var resetterLanguage string
```

BigFab committed:
Apply suggestions from code review

Co-authored-by: yvon-dblg <53182545+yvon-dblg@users.noreply.github.com>

fabrice-diabeloop committed:
Add language from request for resend signup

fabrice-diabeloop committed:
Fix typo

ydrisR reviewed (changes requested):
Looks good and it works as expected (tested in the docker stack).
A couple of remarks:
* please bump the version in the changelog
* would be really good to have unit tests to cover this part of the code. if you don't have time right now I'm ok to merge the PR and log a new task in our backlog.

Nas84 commented on api/forgot.go:
```diff
@@ -40,19 +40,27 @@ type (
 // @Failure 422 {object} status.Status "Error when sending the email (probably caused by the mailling service"
 // @Failure 500 {object} status.Status "Error finding the user, message returned:\"Error finding the user\" "
 // @Router /send/forgot/{useremail} [post]
 func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map[string]string) {
```
@BigFab  this function is very long, do you think it is possible extract some part ?

BigFab replied:
we can find many longer functions in the same project, just with 1 eye open ;)
generally speaking Tidepool developers did create long functions like this one in all the projects
I dont want to review this now if OK for you? @Nas84

Nas84 replied:
yeah i saw 🤣  that is why i proposed to start with only this function. it is a very good exercise to do.
if you think it is not the good time, that ok for me. i don't want to block. 
however i think it should not become a habit, because it is never the good time.

Nas84 reviewed (changes requested):
do we have tests for this ?

ydrisR committed:
Spy on mock notifier Send method

Register how the notifier.Send() method is called.
We want to know which subject it was called with.

ydrisR committed:
Add unit tests to cover the new language preference feature on Signup

ydrisR committed:
Add unit tests to cover the new language preference feature on Signup resend

ydrisR committed:
Add unit tests to cover the new language preference feature on forgot password

fabrice-diabeloop committed:
Add some unit tests for languages

ydrisR reviewed (approved):
OK

yvon-dblg reviewed (approved):
LGTM

Nas84 reviewed (approved):
LGTM
